### PR TITLE
ADBDEV-5746: Add an option to limit number of logical segments to execute external table query

### PR DESCRIPTION
**Pull request by RomaZe**
- Add an option `ACTIVE_SEGMENT_COUNT` for readable external tables to limit number of logical segments to execute a query with readable external tables. Fragments will be distributed across only limit number of segments. 
- Add integration tests for fragments distribution;
- Modify test environment adding additional segment host;
- Fix other integration tests because of the changing test environment.

### CODE REVIEW

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/FragmentDistributionTest.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+    private void preparePgSourceTable() throws Exception {
+        postgresSourceTable = new Table(PG_SOURCE_TABLE_NAME, SOURCE_TABLE_FIELDS);
+        postgresSourceTable.setDistributionFields(new String[]{"id"});
+        gpdb.createTableAndVerify(postgresSourceTable);
+        // Generate rows for source table
+        int count = 20;
+        String[][] rows = new String[count][2];
+        for (int i = 0; i < count; i++) {
+            rows[i][0] = String.valueOf(i+1);
```
Comment (iamlapa) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/FragmentDistributionTest.java]: formatting

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java [hunk 2]
```diff
@@ -131,10 +133,13 @@ public void testPredicateInPostgres() throws Exception {
...
-        String result = grepLog(POSTGRES_SEGMENT_LOG_GREP_COMMAND);
-        assertEquals("1", result);
-        cluster.deleteFileFromNodes(PXF_TEMP_LOG_PATH, false);
+        int result = 0;
+        for (Node pxfNode : pxfNodes) {
+            cluster.copyFromRemoteMachine(pxfNode.getUserName(), pxfNode.getPassword(), pxfNode.getHost(), pxfLogFile, "/tmp/");
+            result += Integer.parseInt(grepLog(POSTGRES_SEGMENT_LOG_GREP_COMMAND));
+            cluster.deleteFileFromNodes(PXF_TEMP_LOG_PATH, false);
+        }
+        assertEquals("Check that filter is present in the log at least on the one segment host", 1, result);
```
Comment (iamlapa) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java:142]: Given that the expected result is 1, I noticed that we're using a `for` loop to handle multiple nodes. Could you clarify if there's a scenario where multiple results could accumulate, justifying the use of the loop? If we anticipate only one result, we might consider simplifying this implementation.

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PxfMetricsTest.java [hunk 3]
```diff
@@ -15,39 +15,49 @@
...
+        Collection<String> pxfHostNames = new ArrayList<>();
         if (cluster instanceof MultiNodeCluster) {
-            Node pxfNode = ((MultiNodeCluster) cluster).getNode(SegmentNode.class, PhdCluster.EnumClusterServices.pxf).get(0);
-            pxfHostName = pxfNode.getHostName();
-            if (StringUtils.isEmpty(pxfNode.getHostName())) {
-                pxfHostName = pxfNode.getHost();
+            List<Node> pxfNodes = ((MultiNodeCluster) cluster).getNode(SegmentNode.class, PhdCluster.EnumClusterServices.pxf);
+            for (Node pxfNode : pxfNodes) {
+                String pxfHostName = pxfNode.getHostName();
+                if (StringUtils.isEmpty(pxfNode.getHostName())) {
```
Comment (iamlapa) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/PxfMetricsTest.java]:
```java
pxfHostNames.add(Optional.ofNullable(pxfNode.getHostName())
                        .filter(StringUtils::isNotEmpty)
                        .orElse(pxfNode.getHost()));
```

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PxfReloadTest.java [hunk 4]
```diff
@@ -298,4 +346,11 @@ private int countArrayListsWithField(List<List<String>> listOfLists, String sear
         }
         return count;
     }
+
+    private String grepLog(String command) throws ShellCommandErrorException, IOException {
+        cluster.runCommand(command);
+        String result = cluster.getLastCmdResult();
+        String[] results = result.split("\r\n");
```
Comment (iamlapa) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/PxfReloadTest.java]: I noticed that you're using a hard-coded line separator (`\r\n`) in the method. To enhance the portability of our code across different operating systems, I suggest using `System.getProperty("line.separator")` to obtain the system-specific line separator instead. This change would help us handle line breaks consistently whether our application is running on Windows, Linux, or any other OS.

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/FragmentDistributionTest.java (hunk 1 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/FragmentDistributionTest.java]: Fixed.

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java (hunk 2 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/PushdownPredicateInTest.java:142]: This test uses readable external table based on the jdbc profile. By default, this profile uses just only 1 logical segment to execute query. Now, we have 2 segment hosts with 3 logical segments on each and we cannot predict where the query will be executed. So, we need to check logs on each segment host and find this record.

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PxfMetricsTest.java (hunk 3 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/PxfMetricsTest.java]: Refactored

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/PxfReloadTest.java (hunk 4 above)
Comment (RomaZe) [reply automation/src/test/java/org/greenplum/pxf/automation/arenadata/PxfReloadTest.java]: Refactored and moved to the util method as we used it in several tests.

**Review by iamlapa — APPROVED** [review]

File: automation/src/test/java/org/greenplum/pxf/automation/arenadata/FragmentDistributionTest.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            rows[i][0] = String.valueOf(i + 1);
+            rows[i][1] = "text" + (i + 1);
+        }
+        dataTable = new Table("dataTable", SOURCE_TABLE_FIELDS);
+        dataTable.addRows(rows);
+        gpdb.insertData(dataTable, postgresSourceTable);
+    }
+
+    private void prepareHdfsSourceTable() throws Exception {
+        ReadableExternalTable hdfsSourceTable = TableFactory.getPxfHcfsWritableTable(
```
Comment (xardazzzzzz) [inline automation/src/test/java/org/greenplum/pxf/automation/arenadata/FragmentDistributionTest.java:91]: getPxfHcfsWritableTable -> getPxfHdfsWritableTable

**Review by xardazzzzzz — APPROVED** [review]

**Review by iamlapa — APPROVED** [review]

**Review by iamlapa — APPROVED** [review]

**Pull request merged**